### ISSUE TITLE
Terminal hangs and errs with bad fd

### ISSUE THREAD

**Issue by ellisonbg** [takluyver/terminado#13]
When I leave a Terminal open in the IPython Notebook for a long period of time, it eventually stops responding and shows a traceback on the server side:

```python
[E 2015-04-06 18:40:47.096 ahenriks ioloop:588] Exception in callback (36, <function wrap.<locals>.null_wrapper at 0x7f8869f13f28>)
    Traceback (most recent call last):
      File "/usr/local/lib/python3.4/dist-packages/tornado/ioloop.py", line 840, in start
        handler_func(fd_obj, events)
      File "/usr/local/lib/python3.4/dist-packages/tornado/stack_context.py", line 275, in null_wrapper
        return fn(*args, **kwargs)
      File "/usr/local/lib/python3.4/dist-packages/terminado/management.py", line 187, in pty_read
        s = ptywclients.ptyproc.read(65536)
      File "/usr/local/lib/python3.4/dist-packages/ptyprocess/ptyprocess.py", line 803, in read
        b = super(PtyProcessUnicode, self).read(size)
      File "/usr/local/lib/python3.4/dist-packages/ptyprocess/ptyprocess.py", line 507, in read
        s = self.fileobj.read(size)
    OSError: [Errno 9] Bad file descriptor
```

The odd thing is that these are raised at the rate of a hundreds/thousands a minute. Not sure if this is an issue with terminado or ptyprocess.

**Comment by minrk** [comment]
My guess is the fileobj has been closed or some such, and isn't being checked before reading. You get EBADF when trying to open a nonexistent FD.

**Comment by takluyver** [comment]
Could this be related to App Nap? I will try to leave a terminal open for some time to see if this happens on Linux. What sort of time period seems to cause it?

The rate of raising isn't terribly surprising - if the code doesn't catch the error and remove the reader from the event loop, I imagine that every iteration of the event loop will retry the read.

**Comment by minrk** [comment]
I'm pretty sure this is already linux, since the path contains `dist-packages`.

**Comment by takluyver** [comment]
Good point.

If the child closes the tty, the parent should get EIO when it tries to read. I *think* EBADF means that something has closed the fd from the parent (server) side without going through terminado/pexpect.

**Comment by ellisonbg** [comment]
Yes, this is a linux system.

On Tue, Apr 7, 2015 at 8:03 PM, Thomas Kluyver <notifications@github.com>
wrote:

> Good point.
>
> If the child closes the tty, the parent should get EIO when it tries to
> read. I *think* EBADF means that something has closed the fd from the
> parent (server) side without going through terminado/pexpect.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/takluyver/terminado/issues/13#issuecomment-90790242>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by ellisonbg** [comment]
@minrk @takluyver I am now seeing this in 4.0.6 very regularly, without leaving the terminal idle for a while (again on Linux). 

Any thoughts on a fix for this?

**Comment by takluyver** [comment]
Is there anything that will reproduce it reliably? I've just had a quick fiddle with some terminals and didn't immediately reproduce it.

**Comment by ellisonbg** [comment]
I am giving you an account on the server that has it running and will send you details on a private email

**Comment by ellisonbg** [comment]
The way I am seeing it is this:

* Tail the log file to watch
* Open a terminal, close it using `$ exit`
* Repeat a few times and eventually it will happen



On Tue, Dec 8, 2015 at 10:38 AM, Thomas Kluyver <notifications@github.com>
wrote:

> Is there anything that will reproduce it reliably? I've just had a quick
> fiddle with some terminals and didn't immediately reproduce it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/takluyver/terminado/issues/13#issuecomment-162974652>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by ellisonbg** [comment]
Sometimes, when terminals open, they don't work at all and I see this in the JS console:

![screen shot 2015-12-08 at 11 20 31 am](https://cloud.githubusercontent.com/assets/27600/11665738/c90215de-9d9d-11e5-9c97-8503de5edd8e.png)

In that case though, no bad FD traceback in the logs.

**Comment by ellisonbg** [comment]
Also, when the terminals get into that state, I can't shut them down in the dashboard.

**Comment by ellisonbg** [comment]
Take that back, I do see the tracebacks when I see the JS console error message. I don't know if it matters, but I typically see the error the 3rd time I open/close a terminal.

**Comment by ellisonbg** [comment]
In particular, it looks like the problem comes from open/close cycling a terminal with the same name. Almost like the FDs from previous terminal sessions are still being used.

**Comment by takluyver** [comment]
There can be some weird effects if fds are closed without their corresponding objects being destroyed, because fds are readily reused, and the objects that think they own them can't tell if the fd refers to the same thing they originally opened. Thanks for the hints, I'll investigate.

**Comment by ellisonbg** [comment]
I am chasing it right now as well - we are definitely getting OSErrors in
the read loop. I think somehow old terminals are not always getting cleaned
up fully.

On Tue, Dec 8, 2015 at 11:42 AM, Thomas Kluyver <notifications@github.com>
wrote:

> There can be some weird effects if fds are closed without their
> corresponding objects being destroyed, because fds are readily reused, and
> the objects that think they own them can't tell if the fd refers to the
> same thing they originally opened. Thanks for the hints, I'll investigate.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/takluyver/terminado/issues/13#issuecomment-162993105>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by ellisonbg** [comment]
When there is a bad fd, the first place I see an exception is when
resize_to_smallest is called at L129 of websocket.py. That makes sense
as it is the first place the fd is written to upon a connection.

On Tue, Dec 8, 2015 at 12:19 PM, Brian Granger <ellisonbg@gmail.com> wrote:
> I am chasing it right now as well - we are definitely getting OSErrors in
> the read loop. I think somehow old terminals are not always getting cleaned
> up fully.
>
> On Tue, Dec 8, 2015 at 11:42 AM, Thomas Kluyver <notifications@github.com>
> wrote:
>>
>> There can be some weird effects if fds are closed without their
>> corresponding objects being destroyed, because fds are readily reused, and
>> the objects that think they own them can't tell if the fd refers to the same
>> thing they originally opened. Thanks for the hints, I'll investigate.
>>
>> —
>> Reply to this email directly or view it on GitHub.
>
>
>
>
> --
> Brian E. Granger
> Associate Professor of Physics and Data Science
> Cal Poly State University, San Luis Obispo
> @ellisonbg on Twitter and GitHub
> bgranger@calpoly.edu and ellisonbg@gmail.com



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by zyzhu2000** [comment]
@ellisonbg , are you using nginx or some proxy? It could be a connection timeout problem. The connection is eventually closed after some time of no activity. I had been suffering a similar problem to yours. My latest pull request fixed my problem.

**Comment by ellisonbg** [comment]
I am using nginx. The recent commit seems to have fixed the problem though.

On Fri, Feb 5, 2016 at 10:14 PM, zyzhu2000 <notifications@github.com> wrote:

> @ellisonbg <https://github.com/ellisonbg> , are you using nginx or some
> proxy? It could be a connection timeout problem. The connection is
> eventually closed after some time of no activity. I had been suffering a
> similar problem to yours. My latest pull request fixed my problem.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/takluyver/terminado/issues/13#issuecomment-180702130>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by takluyver** [comment]
This was fixed by #24
